File: 02-langchain-fastapi-vacation-recommender/vacatify/routes/vacation.py
import uuid

from fastapi import APIRouter, Request, BackgroundTasks, HTTPException

from vacatify.schemas import (
    GenerateVacationIdeaResponse,
    GetVacationIdeaResponse,
    GenerateVacationIdeaRequest,
)

from vacatify.chains.vacation import generate_vacation_idea_chain, vacations

"""
The `vacation_router` is the router that will be used
by the main application. All of the endpoints on the router
will therefore be added to the application so we can 
begin to leverage them!
"""
vacation_router = APIRouter(prefix="/vacation")


"""
The first endpoint we will add is a POST endpoint at
`http://the-api-ip:the-api-port/vacation/`. The main purposes
of this endpoint are to:

1. Start a new background task to create a vacation idea for you
2. Report the created ID and progress back to the user

We can also see that this endpoint takes, as input, a request
of type `GenerateVacationIdeaRequest` and returns a response
of type `GenerateVacationIdeaResponse`. These data types were
defined above in our schemas section. Thankfully, fastAPI does
all of the required serialization both to and from the API!

This method also leverages a FastAPI notion called BackgroundTasks.
We can define background tasks to be run after returning a response.
This is useful because, at time, Langchain might take a while to run.
You don't want your clients to have to handle huge timeouts beacuse that
might indicate other issues. By scheduling these as background tasks, we can
say "hey client, the work has started. Check back later :)"

So, let's look more in depth at `generate_vacation()`. First, 
we generate a UUID for the idea. Then, we submit the chain
with all of the parameters from our user to our background task.
The observant reader will see the function `generate_vacation_idea_chain`.
This is the function that will run langchain and update our vacation
database. We will talk in more depth about it in a following section.
Finally, we return the response with the ID and the completed 
flag set to false.
"""


@vacation_router.post(
    "/",
    summary="Generate a vacation idea.",
    responses={
        201: {"description": "Successfully initiated task."},
    },
)
async def generate_vacation(
    r: GenerateVacationIdeaRequest, background_tasks: BackgroundTasks
) -> GenerateVacationIdeaResponse:
    """Initiates a vacation generation for you."""

    idea_id = uuid.uuid4()
    background_tasks.add_task(
        generate_vacation_idea_chain,
        idea_id,
        r.favorite_season,
        r.hobbies,
        r.budget,
    )
    return GenerateVacationIdeaResponse(id=idea_id, completed=False)


"""
Our second endpoint is a GET endpoint at
`http://the-api-ip:the-api-port/vacation/<id>`. The main
purpose of this endpoint is going to be use to either 
poll/query/read the idea that was created from the generation above.

We can see that is accepts a UUID id as a parameter and returns a 
`GetVacationIdeaResponse` which was again defined in our schema section
above.

We can see that this endpoint just looks to see if a vacation ID matching
the requested ID is present in our vacation "database". I say database in
quotes because this is just a dictionary that is shared across the system.
Ideally, this would be some more persistent/stable/scalable form of storage
but, for the purpose of this conversation, a dictionary is perfect.

If we know the ID that the user is requesting, we can put all the relevant
data into a `GetVacationIdeaResponse` and return it to the user. Otherwise,
we throw a 404.
"""


@vacation_router.get(
    "/{id}",
    summary="Get the generated a vacation idea.",
    responses={
        200: {"description": "Successfully fetched vacation."},
        404: {"description": "Vacation not found."},
    },
)
async def get_vacation(r: Request, id: uuid.UUID) -> GetVacationIdeaResponse:
    """Returns the vacation generation for you."""
    if id in vacations:
        vacay = vacations[id]
        return GetVacationIdeaResponse(
            id=vacay.id, completed=vacay.completed, idea=vacay.idea
        )
    raise HTTPException(status_code=404, detail="ID not found")

File: 07-golang-gin-langchain/routes/vacation.go
package routes

import (
	"net/http"

	"github.com/afoley587/52-weeks-of-projects/07-golang-gin-langchain/chains"
	"github.com/google/uuid"

	"github.com/gin-gonic/gin"
)

/*

Now that our request and response schemas are out
of the way, we can talk about the router.

The `GetVacationRouter` function takes a gin router
as input and adds a new router grouping to it with a
path prefix of `/vacation`. So any endpoints we add to
the router will get that `/vacation` prefix. We then
add two endpoints:

1. POST `/create` to create a new vacation idea for us
2. GET `/:id` which gets an idea by ID

From a high level, the `/create` endpoint will kick
off a goroutine which invokes langchain and openAI. It
will return a `GenerateVacationIdeaResponse` to the caller
so that they can check back on the status of it later. They
check the status of that idea with the `/:id` endpoint. This
will return a `GetVacationIdeaResponse`. If the idea is finished
generating, it will contain an id, an idea, and the completed flag
will be set to true. Otherwise, it will contain an id, an empty idea,
and the completed flag will be set to false.

So, in more depth, what happens when we send a POST to `/create`? First,
it will validate the request, i.e. making sure that the request that
was sent in is valid. If it is, it will call `generateVacation` with
the deserialized request. `generateVacation` will generate a new UUID
for the idea and invoke the langchain chain in a goroutine with
`go chains.GeneateVacationIdeaChange(id, r.Budget, r.FavoriteSeason, r.Hobbies)`.
We will talk about what this chain is doing in the next section but, for now,
let's just say that it goes and generates a vacation idea with the given parameters.
After it starts that chain, it returns a `GenerateVacationIdeaResponse` to the
caller with the ID field set. We should also note that it's important to us to
put this on a goroutine because we want our responses to be snappy. Langchain
might take a few seconds to generate an actual idea so we don't want clients
to register things like timeouts when waiting for a response.

How about the GET to `/:id`? Well, this one's a bit simpler. It also first
validates the request by making sure that the ID is a valid UUID. It then goes
and tries to query the in-memory DB (a map) that we keep for vacation idea
book-keeping. If the ID doesn't exist, we will return a 404 to the caller.
If it does, we will turn the `Vacation` object into a `GetVacationIdeaResponse`
containing all the releveant data (ID, idea, and whether it's finished or not).
*/

func generateVacation(r GenerateVacationIdeaRequest) GenerateVacationIdeaResponse {
	// First, generate a new UUID for the idea
	id := uuid.New()

	// Then invoke the GeneateVacationIdeaChange method of the chains package
	// passing through all of the parameters from the user
	go chains.GeneateVacationIdeaChange(id, r.Budget, r.FavoriteSeason, r.Hobbies)
	return GenerateVacationIdeaResponse{Id: id, Completed: false}
}

func getVacation(id uuid.UUID) (GetVacationIdeaResponse, error) {
	// Search the chains database for the ID requested by the user
	v, err := chains.GetVacationFromDb(id)

	// If the ID didn't exist, handle the error
	if err != nil {
		return GetVacationIdeaResponse{}, err
	}

	// Otherwise, return the vacation idea to the caller
	return GetVacationIdeaResponse{Id: v.Id, Completed: v.Completed, Idea: v.Idea}, nil
}

func GetVacationRouter(router *gin.Engine) *gin.Engine {

	// Add a new router group to the gin router
	registrationRoutes := router.Group("/vacation")

	// Handle the POST to /create
	registrationRoutes.POST("/create", func(c *gin.Context) {
		var req GenerateVacationIdeaRequest
		err := c.BindJSON(&req)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"message": "Bad Request",
			})
		} else {
			c.JSON(http.StatusOK, generateVacation(req))
		}
	})

	// Handle the GET to /:id
	registrationRoutes.GET("/:id", func(c *gin.Context) {
		id, err := uuid.Parse(c.Param("id"))

		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"message": "Bad Request",
			})
		} else {
			resp, err := getVacation(id)
			if err != nil {
				c.JSON(http.StatusNotFound, gin.H{
					"message": "Id Not Found",
				})
			} else {
				c.JSON(http.StatusOK, resp)
			}
		}
	})

	// Return the updated router
	return router
}
